internal/controllers/clusterextension_controller.go

```diff
@@ -112,88 +179,244 @@ func checkForUnexpectedFieldChange(a, b ocv1alpha1.ClusterExtension) bool {
-	// Lookup the bundle that corresponds to the ClusterExtension's desired package.
-	bundle, err := r.resolve(ctx, ext)
+	l := log.FromContext(ctx).WithName("operator-controller")
+	// run resolution
+	bundle, err := r.resolve(ctx, *ext)
 	if err != nil {
-		ext.Status.InstalledBundle = nil
-		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
-		ext.Status.ResolvedBundle = nil
-		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+		l.V(1).Info("bundle resolve error", "error", err)
+		return r.handleResolutionErrors(ext, err)
+	}
+
+	if err := r.validateBundle(bundle); err != nil {
+		setInstalledStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as installation has failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
 
-		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+	bundleVersion, err := bundle.Version()
+	if err != nil {
+		ext.Status.ResolvedBundle = nil
+		ext.Status.InstalledBundle = nil
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", ocv1alpha1.ReasonResolutionFailed, err), ext.GetGeneration())
+		setInstalledStatusConditionFailed(&ext.Status.Conditions, fmt.Sprintf("%s:%v", ocv1alpha1.ReasonInstallationFailed, err), ext.Generation)
 		return ctrl.Result{}, err
 	}
 
-	// Now we can set the Resolved Condition, and the resolvedBundleSource field to the bundle.Image value.
 	ext.Status.ResolvedBundle = bundleMetadataFor(bundle)
 	setResolvedStatusConditionSuccess(&ext.Status.Conditions, fmt.Sprintf("resolved to %q", bundle.Image), ext.GetGeneration())
 
-	// TODO: Question - Should we set the deprecation statuses after we have successfully resolved instead of after a successful installation?
+	// Generate a BundleDeployment from the ClusterExtension to Unpack.
+	// Note: The BundleDeployment here is not a k8s API, its a simple Go struct which
+	// necessary embedded values.
+	bd := r.generateBundleDeploymentForUnpack(bundle.Image, ext)
+	unpackResult, err := r.Unpacker.Unpack(ctx, bd)
+	if err != nil {
+		return ctrl.Result{}, updateStatusUnpackFailing(&ext.Status, err)
+	}
+
+	switch unpackResult.State {
+	case rukpaksource.StatePending:
+		updateStatusUnpackPending(&ext.Status, unpackResult)
+		// There must be a limit to number of entries if status is stuck at
+		// unpack pending.
+		return ctrl.Result{}, nil
+	case rukpaksource.StateUnpacking:
+		updateStatusUnpacking(&ext.Status, unpackResult)
+		return ctrl.Result{}, nil
+	case rukpaksource.StateUnpacked:
+		// TODO: Add finalizer to clean the stored bundles, after https://github.com/operator-framework/rukpak/pull/897
+		// merges.
```

Comment:
Note for other reviewers, this looks like it will be addressed as a follow-up. Looks like https://github.com/operator-framework/operator-controller/issues/868 is the corresponding issue to track this.

@varshaprasad96 (and others) please feel free to correct me if I am incorrect.